### PR TITLE
167190491 User Login

### PR DESCRIPTION
**Pull request by JuwonAbiola**
#### What does this PR do?
- Enable user login functionality on the Authors Haven app

#### Description of Task proposed in this pull request?
- Build and style the login page
- Consume the login endpoint
- Write component and e2e test

#### How should this be manually tested (Quality Assurance)?
- Run `npm run start:dev'
- A browser tab will launch then navigate to /login
- Input a valid email or username and password

#### What are the relevant pivotal tracker stories?
- [167190491](https://www.pivotaltracker.com/story/show/167190491)

#### Any background context you want to add (Operations Impact)?
- Run npm install to download the project's dependencies

#### What I have learned working on this feature: 
- I learned how to map state to props

#### Screenshots: 
-

### CODE REVIEW

File: src/styles/index.scss [hunk 1]
```diff
@@ -140,3 +147,30 @@ input[type='password'] {
...
+        &::placeholder {
+          font-weight: 200;
+          font-size: 0.9em;
+        }
+    }
+
+    .has-validation-error {
+        border-color:$redColor !important;
+    }
+
```
Comment (dinobi) [inline src/styles/index.scss]: add new line

File: src/reducers/user/login.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * @param {action} action - action creator
+ * @param {action.types} action.types - action type
+ *
+ * @return {object} - new state object
+ */
+const loginReducer = (state = initialState, action) => {
+  switch (action.type) {
+    case actionTypes.LOGIN_REQUEST:
+      return {
+        state, loggingIn: true, isLoading: true
```
Comment (dinobi) [inline src/reducers/user/login.js]: what is `loggingIn`?

File: src/reducers/user/login.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      };
+    case actionTypes.LOGIN_SUCCESS:
+      return {
+        state,
+        user: action,
+        isLoading: false
+      };
+    case actionTypes.LOGIN_FAILURE:
+      return {
```
Comment (dinobi) [inline src/reducers/user/login.js:30]: why don't you maintain consistency with your object structure?

File: src/components/Login/login.scss [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+ span > a {
+    margin: 5px
+  }
+.center{
+    margin: 0 auto;
+    text-align: center;
+}
+.input-validation-error {
+    font-size: 0.95em;
+    color:red !important;
```
Comment (abdulfatai360) [inline src/components/Login/login.scss]: @jhayX You didn't remember to replace the `red` value with the `$redColor` variable you added in the `variables.scss` file below.

File: src/components/Login/login.jsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+               <p className="center">Forgot Password?</p>
+             </Form>
+             <div className="form__footer">
+               <span>
+                Continue with:
+                 <img className="facebook-icon" src={FacebookIcon} alt="facebook" />
+                 <img className="google-icon" src={GoogleIcon} alt="google" />
+                 <img className="twitter-icon" src={TwitterIcon} alt="twitter" />
+               </span>
+               <span>Don't have an account? login</span>
```
Comment (abdulfatai360) [inline src/components/Login/login.jsx]: @jhayX I believe we expect this link should redirect users to the signup page, right? You could make use of the `Link` component of `react-router-dom` package to achieve this.

File: src/components/Login/login.jsx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+               {error && (<p className="center input-validation-error">{errors.server}</p>)}
+               <div className="form-field">
+                 <Button
+                   className="btn"
+                   title="Please fill up the form"
+                   label={isLoading ? 'Please wait...' : 'Submit'}
+                   disabled={isLoading}
+                 />
+               </div>
+               <p className="center">Forgot Password?</p>
```
Comment (abdulfatai360) [inline src/components/Login/login.jsx]: @jhayX Could you also kindly redirect users to the `request password` page when they click on this? Kindly take advantage of `Link` component of `react-router-dom` to achieve this

File: src/components/Login/login.jsx (hunk 5 above)
Comment (ayodejiAA) [reply src/components/Login/login.jsx]: @jhayX In addition, Why should you redirect a user to login form again? Please, let's change this to signup.

File: src/components/Login/login.spec.js [hunk 7]
```diff
@@ -1,12 +1,92 @@
...
+    expect(FormComponent.dive().find('Button')).toHaveLength(1);
+    expect(button.props().label).toEqual('Submit')
+
+  });
+});
+
+describe('Login Component interaction', () => {
+  it('updates its state when input fields change', () => {
+    const { wrapper } = setup();
+    const classInstance = wrapper.instance();
```
Comment (dinobi) [inline src/components/Login/login.spec.js]: please use the proper variable name here

File: src/components/Login/login.jsx [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    userLogin: '',
+    password: '',
+    errors: {
+      userLogin: '',
+      password: '',
+      server: ''
+    }
+  };
+
+  handleChange=(e) => {
```
Comment (dinobi) [inline src/components/Login/login.jsx]: put spaces between operators

**Review by dinobi — APPROVED** [review]

**Pull request merged**